### ISSUE TITLE
🚀 Feature Request: support for top level await?

### ISSUE THREAD

**Issue by renhiyama** [cloudflare/workers-sdk#2029]
### Describe the solution

Some libraries already use top level `await`. And I think you should allow this too. I dont see any problems with this, and if there is any, can be easily bypassed with an async function, but it harms code readability (dont want my whole file's code to go inside an async function). So can we please consider this?
```js
✘ [ERROR] Build failed with 23 errors:

  ../imports/URLImport.js:4:11: ERROR: Top-level await is not available in the configured target
  environment ("es2020")
  ../imports/URLImport.js:5:24: ERROR: Top-level await is not available in the configured target
  environment ("es2020")
  ../imports/URLImport.js:6:9: ERROR: Top-level await is not available in the configured target
  environment ("es2020")
  ../imports/URLImport.js:9:19: ERROR: Top-level await is not available in the configured target
  environment ("es2020")
  ../imports/URLImportInstaller.js:4:9: ERROR: Top-level await is not available in the configured
  target environment ("es2020")
  ...


If you think this is a bug then please create an issue at https://github.com/cloudflare/wrangler2/issues/new/choose
```

**Comment by caass** [comment]
There may be more ways to avoid using a top-level await within a worker -- for example, within a request handler, you can specify pass a promise to [`ctx.waitUntil`](https://developers.cloudflare.com/workers/runtime-apis/fetch-event/#waituntil). I understand, though, that this harms code readability as well.

There's an [open issue](https://github.com/evanw/esbuild/issues/253) with the bundler we use internally to support top-level await, and from skimming quickly it looks like while it would be [technically possible](https://github.com/evanw/esbuild/issues/253#issuecomment-1017410747) for us to support TLA, the actual implementation [is not ideal](https://github.com/evanw/esbuild/issues/253#issuecomment-950260550).

I'll keep this issue open for now as a way of tracking https://github.com/evanw/esbuild/issues/253, so that once support is added in esbuild we can enable it within wrangler -- but given that the last update was about a year ago, I don't anticipate this being added any time soon.

**Comment by IgorMinar** [comment]
hmm... the approach I proposed above might trigger "The script will never generate a response." error described in https://github.com/cloudflare/workerd/issues/210 if the runtime is hit by two requests while awaiting the initialization.

**Comment by IgorMinar** [comment]
I tried various things (especially using context.waitUntil) to make the top level await work, but nothing worked reliably, so I ended up resorting to the only thing that does seem to work reliably:

`main.ts`:
```ts
export default {
  async fetch(request, environment, context) {
    const myWorker = (await import('./my-worker')).default;
    return myWorker.fetch(request, environment, context);
  }
```

`my-worker.ts`:
```ts
const workerInit = Promise.resolve('yay, it works!');

export default {
  async fetch(request, environment, context) {
    await workerInit;
    ...
  }
}
```

**Comment by IgorMinar** [comment]
I believe that this is now resolved with fix for #2716, there is still an issue related to https://github.com/cloudflare/workerd/issues/389 and https://github.com/cloudflare/workerd/issues/402 but those should be tracked separately.

**Comment by IgorMinar** [comment]
to clarify, I verified that top-level `await` works as expected except for that odd issue with esbuild, which I think stems from awaiting WASM events.
